### ISSUE TITLE
Auto logging/scalar detection stopped working in new version

### ISSUE THREAD

**Issue by ndalton12** [allegroai/clearml#322]
I had previously been using clearml version 0.17.4 with pytorch and pytorch lightning both at latest release. I recently updated clearml to the latest version 0.17.5. However, it appears that the auto logging/iteration detection does not work now. Before, the iteration and scalar detection (via tensorboard file) would be picked up automatically. However, now it only works once in a while or not at all, usually reporting no scalars and no iterations. 

My script does a lot of caching up front before iterations actually start, takes maybe 3 minutes? Perhaps this is part of the issue. I can't share the code, however, but other than that it is standard pytorch/pytorch lightning. 

Reverting to 0.17.4 seems to fix the issue.

**Comment by bmartinn** [comment]
Hi @ndalton12 
Yes, in 0.17.5 we rewrote parts of the background reporting (logging), basically switching from background thread to background sub-process.
Regrading the logging issue, a few setup questions:
- Are you running with an agent, or is it manually executed ?
- What's OS ? Python version ? 
- When you say "auto logging/iteration detection does not work", you mean there are no scalars at all?
- What is " once in a while" ? is it in the same process / Task, or multiple Tasks ?
- "caching up front before iterations actually start" , do you see any other reports while caching (e.g. console outputs etc)

**Comment by ndalton12** [comment]
Hi @bmartinn, 

* It is manually executed.
* OS is Ubuntu 20.04.2 LTS, Python version is 3.8.5
* Yes, only the monitoring scalars show up, none of the ones actually put into the tensorboard log files.
* When I manually execute, it makes a new task. It looks like clearml did manage to auto detect the iterations once or twice over maybe 6 runs or so? 
* Yes, the caching generates output progress for the caching. Clearml generates outputs to the console while starting up, as expected.

```
ClearML Task: created new task id=c84611ac14db4f57b7d1fdaf0cc80cc8
2021-03-22 15:38:28,748 - clearml.Task - INFO - No repository found, storing script code instead
ClearML results page: http://10.1.10.240:8080/projects/cccf1c05ba3f4a8aa9d3113f2d5d6e78/experiments/c84611ac14db4f57b7d1fdaf0cc80cc8/output/log
2021-03-22 15:38:28
GPU available: True, used: True
TPU available: None, using: 0 TPU cores
Using native 16bit precision.
2021-03-22 15:38:30
Using native 16bit precision.
Global seed set to 1337
initializing ddp: GLOBAL_RANK: 1, MEMBER: 2/2
2021-03-22 15:38:31
Global seed set to 1337
initializing ddp: GLOBAL_RANK: 0, MEMBER: 1/2
2021-03-22 15:38:33
| Name | Type | Params
---------------------------------------------------
0 | model | ResNet | 23.5 M
1 | criterion | CrossEntropyLoss | 0
2 | train_metrics | MetricCollection | 0
3 | val_metrics | MetricCollection | 0
4 | test_metrics | MetricCollection | 0
---------------------------------------------------
4.1 K Trainable params
23.5 M Non-trainable params
23.5 M Total params
94.049 Total estimated model params size (MB)
2021-03-22 15:38:55
Loading Samples: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 308/308 [00:21<00:00, 14.09samples/s]
```

More output is actually put to the console after this caching is done (ie the actual training from pytorch lightning), but clearml fails to log any of it after this point. The web console does show that the program is still running though, so it hasn't lost all connection to the running program.

**Comment by bmartinn** [comment]
@ndalton12 a few followup questions:
What's the tensorboard version? and the pytorch-lightining version?
When you say "auto detect the iterations once" , do you mean log the scalars reported to the TB ?
You mentioned "over maybe 6 runs or so..."  are those different executions (i.e. multiple processes executed) or is it in the same process ? 
Are you running a few experiments in parallel on the same machine with the same project/task name ?

**Comment by ndalton12** [comment]
Also, I am using DDP for distributed training, so maybe that is related. But everything worked perfectly in version 0.17.4, including when I used DDP or not.

**Comment by vlfom** [comment]
Exactly the same issue in my case and switching back to v0.17.4 helped.

In addition to that, previously, ClearML was able to detect & report the Jupyter Notebook cell's output in `Results -> Console`.
Might be expected behavior, but I didn't manage to get the same in with v0.17.5.

Specifically what I'm talking about:
```python
from clearml import Task

task = Task.init(...)

# The console output from the command below would get logged
!python train.py
```

**Comment by bmartinn** [comment]
@vlfom are you saying that with 0.17.4 when running inside a Jupyter notebook, bash outputs were logged in the UI?
Or is it that the `!python train.py` is running another script that calls Task.init and that script is logged in the UI?
If it is the second, that actually kind of makes sense as this is similar to the DDP.
How to I reproduce it?

**Comment by bmartinn** [comment]
@vlfom , @ndalton12 any thoughts on how to reproduce it? 
I would love to get it fixed, we are ramping up towards a new major version ;)

**Comment by ndalton12** [comment]
@bmartinn Let me see if I can create a reproduction example.

**Comment by ndalton12** [comment]
```
import os
import time
from argparse import ArgumentParser

import pytorch_lightning as pl
from clearml import Task
from pytorch_lightning.core.lightning import LightningModule
from torch import nn
from torch.nn import functional as F
from torch.optim import Adam
from torch.utils.data import DataLoader, random_split
from torchvision import transforms
from torchvision.datasets import MNIST


class LitMNIST(LightningModule):

    def __init__(self):
        super().__init__()

        self.layer_1 = nn.Linear(28 * 28, 128)
        self.layer_2 = nn.Linear(128, 256)
        self.layer_3 = nn.Linear(256, 10)

    def forward(self, x):
        batch_size, channels, width, height = x.size()

        # (b, 1, 28, 28) -> (b, 1*28*28)
        x = x.view(batch_size, -1)
        x = self.layer_1(x)
        x = F.relu(x)
        x = self.layer_2(x)
        x = F.relu(x)
        x = self.layer_3(x)

        x = F.log_softmax(x, dim=1)
        return x

    def configure_optimizers(self):
        return Adam(self.parameters(), lr=1e-3)

    def training_step(self, batch, batch_idx):
        x, y = batch
        logits = self(x)
        loss = F.nll_loss(logits, y)
        self.log('my_loss', loss, prog_bar=True, logger=True)
        return loss


class MNISTDataModule(pl.LightningDataModule):
    def __init__(self, batch_size=64):
        super().__init__()
        self.batch_size = batch_size

    def prepare_data(self):
        # download only
        MNIST(os.getcwd(), train=True, download=True, transform=transforms.ToTensor())
        MNIST(os.getcwd(), train=False, download=True, transform=transforms.ToTensor())

    def setup(self, stage):
        # transform
        transform = transforms.Compose([transforms.ToTensor()])
        mnist_train = MNIST(os.getcwd(), train=True, download=False, transform=transform)
        mnist_test = MNIST(os.getcwd(), train=False, download=False, transform=transform)

        # train/val split
        mnist_train, mnist_val = random_split(mnist_train, [55000, 5000])

        # assign to use in dataloaders
        self.train_dataset = mnist_train
        self.val_dataset = mnist_val
        self.test_dataset = mnist_test

    def train_dataloader(self):
        return DataLoader(self.train_dataset, batch_size=self.batch_size)

    def val_dataloader(self):
        return DataLoader(self.val_dataset, batch_size=self.batch_size)

    def test_dataloader(self):
        return DataLoader(self.test_dataset, batch_size=self.batch_size)


def main(args):
    Task.set_credentials(
        api_host='http://localhost:8008',
        web_host='http://localhost:8080',
        files_host='http://localhost:8081'
    )

    task = Task.init(project_name='examples', task_name='Dummy task')

    url = task.get_output_log_web_page()
    assert "10.1.10.240" in url and "clear.ml" not in url and ".com" not in url

    data = MNISTDataModule(batch_size=64)
    model = LitMNIST()

    # Initialize a trainer
    trainer = pl.Trainer(gpus=args.gpus, precision=16, check_val_every_n_epoch=2, accelerator="ddp", max_epochs=10)

    # Train the model ⚡
    trainer.fit(model, datamodule=data)
    trainer.test(model, datamodule=data)


if __name__ == "__main__":
    parser = ArgumentParser()
    parser.add_argument('--gpus', type=int, default=2)

    args = parser.parse_args()
    main(args)
```

I am not able to reproduce the issue easily with a minimal example right now, but I was able to reproduce a potentially related issue. If you run the above code, and manually stop it running by using ctrl+c, the clearml dashboard shows the program as still running. Reverting to 0.17.4 seems to fix the issue. Torch at 1.8.1 and Pytorch lightning at 1.2.7.

Unrelated, but for future versions, would it be possible to show the experiment version number and/or adding a small text description next to each experiment? I like all the features of clearml but I find it hard to sort through experiments just by using the tags.

**Comment by bmartinn** [comment]
Many thanks @ndalton12 !
I'll update here once I'm able to reproduce with the attached code.

**Comment by orgoro** [comment]
I had the same issue - started working when I moved to 0.17.4
tensorboard = 2.4.1

**Comment by bmartinn** [comment]
Thank you again @ndalton12, I tested the toy code with the latest RC, and it seems to solve the issue :)
```bash
pip install clearml==0.17.6rc1
```
If you notice a few warnings when terminating (ctrl-c) mid training, no need to worry, a fix is already on the main branch.
Guys, please let me know if it solves the issue when running your own specific code

**Comment by ndalton12** [comment]
@bmartinn Awesome, thanks! I'll give it a go tomorrow.

I also realized that you can already customize the columns to add more fields, such as descriptions, which is helpful. I guess it never occurred to me to look at the settings haha.

**Comment by ndalton12** [comment]
Hi @bmartinn, Sorry about getting back to you so late, I got a bit sidetracked.

Anyway, I was able to give the new release candidate a try, but it seems there are still some minor problems. Scalars are now being tracked correctly and clearml recognized when I stopped the program. However, I am still getting a message that clearml fails to detect iteration reporting, so it only updates the recorded console output whenever I log metrics by running on the validation set. That or the console output tracking has greatly reduced frequency, but I assume the problem is the former. This is all in comparison to 0.17.4 by the way.

**Comment by tkupek** [comment]
We are having a similar Issue when upgrading from `0.17.5` to `1.0.1` / `1.0.2`.
Scalars and Debug Samples are not updated during training. It's only uploaded when training is finished. Should I open a new Issue or is this related?

**Comment by bmartinn** [comment]
Hi @tkupek 
This all "feels" related, at least for the time being, let's keep the discussion here.
Can you reproduce the issue with the above [code sample](https://github.com/allegroai/clearml/issues/322#issuecomment-817982659)?
What's your python version? What's the OS you are running on?Are you manually running the code, or are you running it remotely with the agent ? What's your `clearml-server` version?

BTW:
@ndalton12 can you verify your issue was solved with `clearml==1.0.2` ?
It seems to work on my machines (Ubuntu 16/18, python 3.x, clearml-server 1.0)

**Comment by tkupek** [comment]
We are using remote execution with the `nvcr.io/nvidia/pytorch:20.12-py3` base docker image with Clearml-Server `allegroai/clearml:latest`.

I can try to reproduce the issue with the code sample above, but will need some time for this.

**Comment by tkupek** [comment]
I re-tested this today with the latest server + client versions and cannot reproduce it.
Seems to be fixed.

**Comment by bmartinn** [comment]
Thanks @tkupek !
@ndalton12 can you also verify the problem is solved with the latest RC
```
pip install clearml==1.0.3rc0
```

**Comment by ndalton12** [comment]
Hi @bmartinn, sorry for the delay. I tried with the latest stable release (1.0.2 I believe, not RC), and the issue is mostly fixed. Logging appears to be working correctly now. Only problem I can see is that sometimes I end my program manually with ctrl+c and the dashboard lists the task as still running. Probably an issue with DDP? When I abort the program from the dashboard instead, both the program and the dashboard work as expected (stopping and listing it as aborted). 

Anyway, since that's a minor problem and this issue seems to be fixed, I will close the issue. Thanks for all the help!

**Comment by orgoro** [comment]
This is still happening to me

Just updated to latest v1.0.2 server + code
There is no scalers update unless I roll back to  `clearml-0.17.5rc5`
Even tried `clearml-1.0.3rc0` no luck

I am starting training using pytorch pre-trained model, everything is fine except there are no updates to the scaler tab

Can someone help? @bmartinn

**Comment by bmartinn** [comment]
Hi @orgoro 
can you test with `clearml-1.0.3rc1` ?
Quick question are you using pytorch distributed ? or pytorch DataLoaders with fork-server ?
